### ISSUE TITLE
`forkr` is likely broken for particle Gibbs 

### ISSUE THREAD

**Issue by yebai** [TuringLang/AdvancedPS.jl#21]
`forkr` is likely broken without explicit dependence on the replaying mechanism in Turing.jl. This is required to correctly implement the reference particle for particle Gibbs. A potential fix is the switch to a simpler replying mechanism by fixing the random seed for the reference particle.

**Comment by devmotion** [comment]
> A potential fix is the switch to a simpler replying mechanism by fixing the random seed for the reference particle.

I thought a bit more about this and noticed that the main problem will be the resampling step - even if the random seed and the random number generator that is passed to the model execution is fixed, then in the resampling step we would like to obtain diverging traces. I.e., it is impossible to fix the model execution in advance by setting a seed initially. Maybe a workaround could be to save the seed in every step and to explicitly reseed the random number generators (which would have to be separate for each particle) in every resampling step.

**Comment by yebai** [comment]
It sounds like a sensible solution. Maybe consider introducing a simple `assume` function, so that we have more control of the model execution than directly calling `rand`?

**Comment by yebai** [comment]
The following package from the `Gen` team might be interesting. 

https://github.com/probcomp/TracedRandom.jl

Maybe we can consider introducing a customised `rand` / `simulate` function, which accepts a **unique** RV name and a distribution as arguments. All RVs can be stored in a standard Julia dictionary (or something similar but more performant) This would allow particle Gibbs (and its ancestor resampling variant) to be supported, without resorting the general-purpose hefty `VarInfo` mechanism from Turing.

**Comment by devmotion** [comment]
I don't think it's a good idea to add yet another signature to `rand` that is not compatible with any other use of `rand` I've seen. In general, it seems the package is mostly type piracy?

Somehow to me it feels the problem of replaying a trace and saving random variables in some structure are two distinct problems (and hence I am not sure if this should be two separate issues). If the samples can be obtained in the order they appear in the model, then in the most basic setting users could map them to variable names manually?

Alternatively, maybe one could perform some IR-rewriting of the model that allows to inject an "context" that saves sampled variables in some data structure.

**Comment by yebai** [comment]
I agree that a clean, minimal implementation is preferred here. We'd like to keep `AdvancedPS`'s dependency footprint small too. 

As for the replaying mechanism, I am thinking about something in the lines of an `TracedRNG <: AbstractRNG`. This `GeneralisedRNG` could record the seed, and the number of randomnesses it has generated. In addition, we can have a `VectorOfTracedRNG <: AbstractRNG` so that it can support the type of operations required by ancestor resampling in particle Gibbs. 

```julia
# Illustration code
mutable struct TracedRNG{T}  <: AbstractRNG where T<:AbstractRNG
   count::Int # Number of randomness it has generated
   seed   # Seed of RNG
   rng::T # Current state of RNG
end
```

`VectorOfTracedRNG <: AbstractRNG` is a vector of `TracedRNG`s. It is helpful in cases where we need to mix and match several streams of randomness, such as when forking the reference particle in particle Gibbs. 

Any thoughts?

**Comment by FredericWantiez** [comment]
To be able to replay the trace given the TracedRNG, I think we need to change the Trace definition to explicitly reference the rng ? That also means we need to change the model definition to take the rng explicitly to be able to keep track of the randomness generated if we mutate the rng in `rand(rng::TracedRNG, ...) `

**Comment by devmotion** [comment]
Yes, I assume the model has to include the `rng` as well, as noted in https://github.com/TuringLang/AdvancedPS.jl/pull/20#issue-535245931. Other algorithms such as MH or HMC were updated when the algorithms in Turing were switched to AbstractMCMC 2, but this was not done for the particle methods.

Maybe it would be useful to also not allow general RNGs but use user-provided RNGs (or the implicitly provided global RNG) only to seed a special class of internally used RNGs (IIRC e.g. StochasticDiffEq does this). Possibly this could allow additional optimizations - e.g., maybe it could be useful to use a RNG design similar to [Jax](https://github.com/google/jax/blob/master/design_notes/prng.md) that is based on a [counter PRNG](http://www.thesalmons.org/john/random123/papers/random123sc11.pdf) and a [splitting model](https://dl.acm.org/doi/10.1145/2578854.2503784)? Intuitively, the splitting model could be useful when resampling particles.

**Comment by yebai** [comment]
Closed by #23
